Add MakeProtective method for creating a Protective MBR.

smoser opened the pull request:
In almost all cases, when creating new partition tables, a user should
create an GPT Table.  GPT requires a Protective MBR.
Add the MakeProtective method to the MBR interface so that someone
can easiliy create a valid Protective MBR.

The size change on the check for ErrorPartitionLastSectorHigh was
required in order to make a partition of the largest uint32 size
that starts at offset 1.  This is how gdisk makes its partition tables
for disks larger than 2TiB.

smoser committed:
Add MakeProtective method for creating a Protective MBR.

In almost all cases, when creating new partition tables, a user should
create an GPT Table.  GPT requires a Protective MBR.
Add the MakeProtective method to the MBR interface so that someone
can easiliy create a valid Protective MBR.

A change is made to Check() to allow TYPE_GPT partitions to
be exactly 2TiB in length even though it is strictly outside the
MBR specification.  All tested gpt partitioning tools create a
partition of length 4294967295 (0xFFFFFFFF).

rekby commented on mbr.go:
```diff
@@ -151,6 +151,34 @@ func (this MBR) IsGPT() bool {
+// MakeProtective - Make this MBR a GPT Protective MBR
+//   sectorSize is either 512 or 4096. diskSize is the size of entire disk in bytes.
+//   https://en.wikipedia.org/wiki/GUID_Partition_Table#Protective_MBR_(LBA_0)
+func (this *MBR) MakeProtective(sectorSize int, diskSize uint64) {
```
I think it need return error if diskSize % sectorSize != 0

smoser replied:
I can do that.

smoser replied:
done.

rekby commented on mbr.go:
```diff
@@ -77,7 +77,7 @@ func (this *MBR) Check() error {
 
 		// Check if partition last sector out of uint32 bounds
-		if uint64(lp.GetLBAStart())+uint64(lp.GetLBALen()) > uint64(0xFFFFFFFF) {
+		if uint64(lp.GetLBAStart())+uint64(lp.GetLBALen()) > uint64(0xFFFFFFFF)+1 {
```
https://en.wikipedia.org/wiki/Master_boot_record
The organization of the partition table in the MBR limits the maximum addressable storage space of a partitioned disk to 2 TiB (2^32 × 512 bytes).

I thing about default behavior must max follow standard.
And I thing about low level library may have tools for out of standard when it need, but it must be manual and when library user know what he do.

I mean about MBR extension (such as +1 byte for gdisk compatible, or disable max byte check (for allow create 2TB partition with 2TB offset) must be enabled by flag.

I suggest to add additional fields to MBR struct for customize checks:
1. AllowableOverflow - for can add a byte for gdisk compatible (and check - if GDisk add exactly one byte always - may add method GDiskCompatible for set value of the field to one).
2. Add method Read to *MBR for read itself with additinal settings.
3. Change current MBR read method - for create empty mbr object, then call mbr.Read(disk) and return &mbr, err

rekby commented on mbr.go:
```diff
@@ -151,6 +151,34 @@ func (this MBR) IsGPT() bool {
+	pt := this.GetPartition(1)
+	pt.SetType(PART_GPT)
+	pt.SetLBAStart(1)
+	pt.SetLBALen(ptLenLBA)
```
As I read in wikipedia -  max  address must be in uint32, else some other 32-tools/bios may have overflow and unexpected behavior.

rekby reviewed (commented):
Hello Scott, thanks for your patch.

I write my things in review

smoser committed:
MakeProtective: error if size is not evenly divisible by sectorSize.

Address Timofey's request to return error in MakeProtective
if diskSize is not evenly divisible by sectorSize.

smoser commented:
@rekby,

I tested with the 3 GPT partition tools that I'm aware of on linux, and they all behave the same way with regard to the Protective MBR with disks larger than 2TiB.

Here are the programs that I used [check-mbr.go.txt](https://github.com/rekby/mbr/files/4201595/check-mbr.go.txt) and [show-me.txt](https://github.com/rekby/mbr/files/4201599/show-me.txt).  The script basically uses parted, sfdisk and sgdisk to write a GPT partition table to a file and then read and checks that with the attached golang code using this repo.

I think its probably better to allow this by default, and create protective mbrs the same way these popular tools do than to be strictly right.  I've adjusted the pull request to only allow the "too big" partition in this scenario (type==gpt, number == 1, len == exactly 2TiB)

See output below:

    === Creating partition on parted.img size 4398046511104 ===
    # showing mbr on parted.img with: go run check-mbr.go parted.img
    ERROR: Last sector too high.
    1 Start:      1 Len: 4294967295 Type: GPT

    === Creating partition on parted.img size 1099511627776 ===
    # showing mbr on parted.img with: go run check-mbr.go parted.img
    1 Start:      1 Len: 2147483647 Type: GPT
    
    === Creating partition on sfdisk.img size 4398046511104 ===
    # showing mbr on sfdisk.img with: go run check-mbr.go sfdisk.img
    ERROR: Last sector too high.
    1 Start:      1 Len: 4294967295 Type: GPT
    
    === Creating partition on sfdisk.img size 1099511627776 ===
    # showing mbr on sfdisk.img with: go run check-mbr.go sfdisk.img
    1 Start:      1 Len: 2147483647 Type: GPT
    
    === Creating partition on sgdisk.img size 4398046511104 ===
    # showing mbr on sgdisk.img with: go run check-mbr.go sgdisk.img
    ERROR: Last sector too high.
    1 Start:      1 Len: 4294967295 Type: GPT
    
    === Creating partition on sgdisk.img size 1099511627776 ===
    # showing mbr on sgdisk.img with: go run check-mbr.go sgdisk.img
    1 Start:      1 Len: 2147483647 Type: GPT

smoser commented:
@rekby,

Do you have any thoughts on this?  My feelings are just that if the spec says one thing, but all implementations do something else, then at least consider the real world as valid.

Possibly related (mentioning here as much for myself later as for you) is https://github.com/karelzak/util-linux/issues/485 .  There is a lot of information there and linked from there.

Thanks.
Scott

smoser commented:
I just read a bit more of the link above (https://github.com/karelzak/util-linux/issues/485).
it seems that fdisk will always write 0xff for the size of the protective gpt partition.  Karel says that this is what parted also does, and per https://thestarman.pcministry.com/asm/mbr/GPT.htm#GPTPT , 

> by always filling this field with 0xFFFFFF, the Windows 7 (or 8) OS is not following the UEFI Specification which states the EndingCHS should be "set to the CHS address of the last logical block on the disk" unless it's too large to be represented

So what I suggest is that we allow any GPT Protective partition to have 0xFFFFFF as its length.  Otherwise this library will identify almost all real-world protective gpt as invalid.

What do you think?

rekby commented:
Hello, sorry for long pause.
I think is must be optional.

For example add enum MakeProtective with options:
Default
DiskSize
MaxSize

Default will one of DiskSize/MaxSize and may change in feature. DiskSize - Create partition by UEFI standard as description, MaxSize - as windows.

And short description for variants in comments.

smoser committed:
Explicitly set protective MBR bootable flag to NonBootable.

If MBR.MakeProtective was called on a MBR that had invalid values
in any of the partition's "bootable" location, then a subsequent
Check() call would fail.

This explicitly sets all partition values to non-bootable.

smoser committed:
Update to add DiskSize/MaxSize input to makeProtective.

smoser commented:
@rekby ,
Please take another look. I just added what I think you were asking for.
Scott

smoser commented:
@rekby just another ping, 4 years later. If this still works and is what you wanted, then one last chance to grab it, or I'll just close it (no hard feelings).  Thanks for your work on mbr.

rekby commented:
Hello, thanks for your improve the library.

Sorry for very long pause
